driver-core/src/main/com/mongodb/client/model/Filters.java

```diff
@@ -485,10 +490,266 @@ public static Bson size(final String fieldName, final int size) {
+     */
+    public static Bson geoWithin(final String fieldName, final Geometry geometry) {
+        return new GeometryOperatorFilter<Geometry>("$geoWithin", fieldName, geometry);
+    }
+
+    /**
+     * Creates a filter that matches all documents containing a field with geospatial data that exists entirely within the specified shape.
+     *
+     * @param fieldName the field name
+     * @param geometry the bounding GeoJSON geometry object
+     * @return the filter
+     * @since 3.1
+     * @mongodb.driver.manual reference/operator/query/geoWithin/ $geoWithin
+     * @mongodb.server.release 2.4
+     */
+    public static Bson geoWithin(final String fieldName, final Bson geometry) {
+        return new GeometryOperatorFilter<Bson>("$geoWithin", fieldName, geometry);
+    }
+
+    /**
+     * Creates a filter that matches all documents containing a field with grid coordinates data that exist entirely within the specified
+     * box.
+     *
+     * @param fieldName   the field name
+     * @param lowerLeftX  the lower left x coordinate of the box
+     * @param lowerLeftY  the lower left y coordinate of the box
+     * @param upperRightX the upper left x coordinate of the box
+     * @param upperRightY the upper left y coordinate of the box
+     * @return the filter
+     * @mongodb.driver.manual reference/operator/query/geoWithin/ $geoWithin
+     * @mongodb.driver.manual reference/operator/query/box/#op._S_box $box
+     * @mongodb.server.release 2.4
+     * @since 3.1
+     */
+    public static Bson geoWithinBox(final String fieldName, final double lowerLeftX, final double lowerLeftY, final double upperRightX,
+                                    final double upperRightY) {
+        BsonDocument box = new BsonDocument("$box",
+                                            new BsonArray(asList(new BsonArray(asList(new BsonDouble(lowerLeftX),
+                                                                                      new BsonDouble(lowerLeftY))),
+                                                                 new BsonArray(asList(new BsonDouble(upperRightX),
+                                                                                      new BsonDouble(upperRightY))))));
+        return new OperatorFilter<BsonDocument>("$geoWithin", fieldName, box);
+    }
+
+    /**
+     * Creates a filter that matches all documents containing a field with grid coordinates data that exist entirely within the specified
+     * polygon.
+     *
+     * @param fieldName the field name
+     * @param points    a list of pairs of x, y coordinates.  Any extra dimensions are ignored
+     * @return the filter
+     * @mongodb.driver.manual reference/operator/query/geoWithin/ $geoWithin
+     * @mongodb.driver.manual reference/operator/query/polygon/#op._S_polygon $polygon
+     * @mongodb.server.release 2.4
+     * @since 3.1
+     */
+    public static Bson geoWithinPolygon(final String fieldName, final List<List<Double>> points) {
```

Comment:
No, this one if for the legacy $polygon: http://docs.mongodb.org/manual/reference/operator/query/polygon/#op._S_polygon.

It's the same in .NET's FilterDefinition, I believe.